### PR TITLE
Claim Undelegated Tokens Unit Test

### PR DESCRIPTION
**Pull request by sampocs**
**Summary**
* Added unit tests for `MsgClaimUndelegatedTokens`
* The ICA calls required a lot of initial state - it was a little too much to mock out for this type of test. I resorted to breaking up the original function into smaller functions so we could run the tests without calling `SubmitTx`

**Testing Plan**
```
make test-unit
```

### CODE REVIEW

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	s.App.StakeibcKeeper.SetEpochTracker(s.Ctx, epochTracker)
+	s.App.RecordsKeeper.SetUserRedemptionRecord(s.Ctx, redemptionRecord)
+
+	connectionEnd := connectiontypes.ConnectionEnd{ClientId: "client-0", Versions: []*connectiontypes.Version{connectiontypes.DefaultIBCVersion}}
+	owner := stakeibc.FormatICAAccountOwner("STRIDE", stakeibc.ICAAccountType_REDEMPTION)
+	portId := "icacontroller-" + owner
+	state := ibctmtypes.NewClientState("STRIDE", ibctmtypes.DefaultTrustLevel, ibctesting.TrustingPeriod, ibctesting.UnbondingPeriod, ibctesting.MaxClockDrift, clienttypes.NewHeight(0, 10), commitmenttypes.GetSDKSpecs(), ibctesting.UpgradePath, false, false)
+	s.App.StakeibcKeeper.IBCKeeper.ClientKeeper.SetClientState(s.Ctx, "client-0", state)
+	s.App.StakeibcKeeper.IBCKeeper.ConnectionKeeper.SetConnection(s.Ctx, connectionId, connectionEnd)
+	s.App.StakeibcKeeper.IBCKeeper.ConnectionKeeper.SetConnection(s.Ctx, connectionId, connectionEnd)
```
Comment (riley-stride) [inline x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go:77]: nice job figuring this stuff out, it's not easy

Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go:77]: is this line duplicated?

Comment (sampocs) [reply x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go:77]: sorry all that stuff was very much in "debugging state". I just wanted to wait to remove it until the tests were done to be safe. Doing it now - will push a commit in a sec!

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	s.Require().True(found)
+	s.Require().False(redemptionRecord.IsClaimable)
+
+	// Confirm pending claims added
+	pendingClaims := s.App.StakeibcKeeper.GetAllPendingClaims(s.Ctx)
+	s.Require().Equal(1, len(pendingClaims))
+	pendingRedemptionRecordIds := pendingClaims[0].UserRedemptionRecordIds
+	s.Require().Equal(1, len(pendingRedemptionRecordIds))
+	s.Require().Equal(tc.redemptionRecordId, pendingRedemptionRecordIds[0])
+	// QUESTION: Anything else to check that I'm missing?
```
Comment (riley-stride) [inline x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go:133]: lgtm!

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+func (s *KeeperTestSuite) TestClaimUndelegatedTokensNoUserRedemptionRecord() {
+	tc := s.SetupClaimUndelegatedTokens()
+	// Remove the user redemption record
+	s.App.RecordsKeeper.RemoveUserRedemptionRecord(s.Ctx, tc.redemptionRecordId)
+
+	_, err := s.msgServer.ClaimUndelegatedTokens(sdk.WrapSDKContext(s.Ctx), &tc.validMsg)
+	expectedErr := "unable to find claimable redemption record: "
+	expectedErr += "could not get user redemption record: GAIA.1.stride_SENDER: user redemption record error"
```
Comment (riley-stride) [inline x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go:143]: clean approach to these error messages, I like it

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	// Confirm pending claims added
+	pendingClaims := s.App.StakeibcKeeper.GetAllPendingClaims(s.Ctx)
+	s.Require().Equal(1, len(pendingClaims))
+	pendingRedemptionRecordIds := pendingClaims[0].UserRedemptionRecordIds
+	s.Require().Equal(1, len(pendingRedemptionRecordIds))
+	s.Require().Equal(tc.redemptionRecordId, pendingRedemptionRecordIds[0])
+	// QUESTION: Anything else to check that I'm missing?
+}
+
+func (s *KeeperTestSuite) TestClaimUndelegatedTokensNoUserRedemptionRecord() {
```
Comment (riley-stride) [inline x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go:136]: Should we add a failure test with bad inputs to submittxs? You know better than me whether that should be tested here or elsewhere

**Review by riley-stride — APPROVED** [review]
Looks good to me! Some nit comments. Great job powering through the channels/ports/connections setup, that stuff can be quite confusing

**Review by riley-stride — APPROVED** [review]
LGTM let's merge! 🚢

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens.go [hunk 5]
```diff
@@ -62,15 +87,20 @@ func (k msgServer) ClaimUndelegatedTokens(goCtx context.Context, msg *types.MsgC
...
 	}
 
+	return &icaTx, nil
+}
+
+func (k Keeper) FlagRedemptionRecordsAsClaimed(ctx sdk.Context, userRedemptionRecord *recordstypes.UserRedemptionRecord, sequence uint64) {
 	// Set isClaimable to false, so that the record can't be claimed again
 	userRedemptionRecord.IsClaimable = false
-	k.RecordsKeeper.SetUserRedemptionRecord(ctx, userRedemptionRecord)
+	k.RecordsKeeper.SetUserRedemptionRecord(ctx, *userRedemptionRecord)
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_claim_undelegated_tokens.go:103]: Why this change out of curiosity?

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens.go [hunk 6]
```diff
@@ -16,6 +16,28 @@ import (
...
+	if err != nil {
+		return nil, sdkerrors.Wrapf(err, "unable to find claimable redemption record")
+	}
+
+	icaTx, err := k.GetRedemptionTransferMsg(ctx, userRedemptionRecord, msg.HostZoneId)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(err, "unable to build redemption transfer message")
+	}
+
+	sequence, err := k.SubmitTxs(ctx, icaTx)
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_claim_undelegated_tokens.go:29]:
```suggestion
	sequence, err := k.SubmitTxs(ctx, icaTx.ConnectionId, icaTx.Messages, icaTx.Account, icaTx.Timeout)
```
In the interest of changing as few lines of code as possible, wdyt of leaving the SubmitTxs function signature unchanged?

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go (hunk 4 above)
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go:119]: Yep agreed

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+import (
+	epochtypes "github.com/Stride-Labs/stride/x/epochs/types"
+	recordtypes "github.com/Stride-Labs/stride/x/records/types"
+	stakeibckeeper "github.com/Stride-Labs/stride/x/stakeibc/keeper"
+	"github.com/Stride-Labs/stride/x/stakeibc/types"
+	stakeibc "github.com/Stride-Labs/stride/x/stakeibc/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	// host "github.com/cosmos/ibc-go/v3/modules/core/24-host"
```
Comment (asalzmann) [inline x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go:12]: nit: rm?

**Review by asalzmann — COMMENTED** [review]
This is awesome! Thank you so much for adding this

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens.go (hunk 6 above)
Comment (sampocs) [reply x/stakeibc/keeper/msg_server_claim_undelegated_tokens.go:29]: yeah good call!

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens.go (hunk 5 above)
Comment (sampocs) [reply x/stakeibc/keeper/msg_server_claim_undelegated_tokens.go:103]: Since the main function was split up, the child functions needed a userRedmptionRecord and I could have either passed the object itself, or a reference. I went with a reference since this seemed more in line with other types of this flavor.
```
# before
func ClaimUndelegatedTokens() {
    userRedemptionRecord := ... // object
    k.RecordsKeeper.SetUserRedemptionRecord(ctx, userRedemptionRecord)
}

# after
func ClaimUndelegatedTokens() {
    userRedemptionRecordPointer := k.GetClaimableRedemptionRecord(ctx, msg) // returns pointer
    k.FlagRedemptionRecordsAsClaimed(ctx, userRedemptionRecordPointer, sequence)
}
func FlagRedemptionRecordsAsClaimed(userRedemptionRecordPointer) {
   k.RecordsKeeper.SetUserRedemptionRecord(ctx, *userRedemptionRecordPointer)
}
```

**Comment by asalzmann** [comment]
Can you run the integration tests on this? Otherwise, lgtm!

**Comment by sampocs** [comment]
Yup all tests succeeded

**Comment by sampocs** [comment]
but will run again after those last couple changes
